File: Java/DoublyLinkedList.java
import java.util.Scanner;

class Node {
    public int data;
    public Node previous;
    public Node next;

    public Node(int data) {
        this.data = data;
        this.previous = null;
        this.next = null;
    }
}

class DoublyLinkedListData {
    public Node head = null;
    public Node tail = null;

    public String insertAtBeginning(int element) {
        Node newNode = new Node(element);
        if (isEmpty()) {
            head = newNode;
            tail = newNode;
            return "Inserted";
        }
        newNode.next = head;
        head.previous = newNode;
        head = newNode;
        return "Inserted";
    }

    public String deleteAtBeginning() {
        if (isEmpty()) return "Linked List is Empty";
        if (head.next == null) {
            head = null;
            tail = null;
            return "Deleted";
        }
        head = head.next;
        head.previous = null;
        return "Deleted";
    }

    public String insertAtEnding(int element) {
        Node newNode = new Node(element);
        if (isEmpty()) {
            head = newNode;
            tail = newNode;
            return "Inserted";
        }
        tail.next = newNode;
        newNode.previous = tail;
        tail = newNode;
        return "Inserted";
    }

    public String deleteAtEnding() {
        if (isEmpty()) return "Linked List is Empty";
        if (head.next == null) {
            head = null;
            tail = null;
            return "Deleted";
        }
        tail = tail.previous;
        tail.next = null;
        return "Deleted";
    }

    public String insertAfterData(int element, int data) {
        if (isEmpty()) return "Linked List is Empty";
        Node newNode = new Node(element);
        Node curr = head;
        while (curr != null) {
            if (curr.data == data) {
                newNode.previous = curr;
                newNode.next = curr.next;
                if (curr.next != null) curr.next.previous = newNode;
                else tail = newNode;
                curr.next = newNode;
                return "Inserted";
            }
            curr = curr.next;
        }
        return "Data not Found";
    }

    public String deleteAfterData(int data) {
        if (isEmpty()) return "Linked List is Empty";
        Node curr = head;
        while (curr.next != null) {
            if (curr.data == data) {
                curr.next = curr.next.next;
                if (curr.next != null) curr.next.previous = curr;
                else tail = curr;
                return "Deleted";
            }
            curr = curr.next;
        }
        if (curr.data == data) return "No element after this data";
        return "Data not Found";
    }

    public String insertBeforeData(int element, int data) {
        if (isEmpty()) return "Linked List is Empty";
        Node newNode = new Node(element);
        Node curr = head;
        while (curr != null) {
            if (curr.data == data) {
                newNode.next = curr;
                newNode.previous = curr.previous;
                if (curr.previous != null) curr.previous.next = newNode;
                else head = newNode;
                curr.previous = newNode;
                return "Inserted";
            }
            curr = curr.next;
        }
        return "Data not Found";
    }

    public String deleteBeforeData(int data) {
        if (isEmpty()) return "Linked List is Empty";
        if (head.data == data) return "No element before this data";
        Node curr = head;
        while (curr != null) {
            if (curr.data == data) {
                curr.previous = curr.previous.previous;
                if (curr.previous != null) curr.previous.next = curr;
                else head = curr;
                return "Deleted";
            }
            curr = curr.next;
        }
        return "Data not Found";
    }

    public String deleteData(int data) {
        if (isEmpty()) return "Linked List is Empty";
        Node curr = head;
        while (curr != null) {
            if (curr.data == data) {
                if (curr.previous != null) curr.previous.next = curr.next;
                else head = curr.next;
                if (curr.next != null) curr.next.previous = curr.previous;
                else tail = curr.previous;
                curr.next = null;
                curr.previous = null;
                return "Deleted";
            }
            curr = curr.next;
        }
        return "Data not Found";
    }

    public void traverse() {
        if (isEmpty()) {
            System.out.println("Linked List is Empty");
            return;
        }
        Node curr = head;
        while (curr != null) {
            System.out.print(curr.data + " ");
            curr = curr.next;
        }
        System.out.println();
    }

    public void reverseTraverse() {
        if (isEmpty()) {
            System.out.println("Linked List is Empty");
            return;
        }
        Node curr = tail;
        while (curr != null) {
            System.out.print(curr.data + " ");
            curr = curr.previous;
        }
        System.out.println();
    }

    public Boolean isEmpty() {
        return head == null;
    }

    public int getLength() {
        int length = 0;
        Node curr = head;
        while (curr != null) {
            length++;
            curr = curr.next;
        }
        return length;
    }

    public int search(int data) {
        if (isEmpty()) {
            System.out.println("Linked List is Empty");
            return -1;
        }
        Node curr = head;
        while (curr != null) {
            if (curr.data == data) return data;
            curr = curr.next;
        }
        return -1;
    }

    public String reverse() { 
        if (head == null) return "Linked List is Empty"; 
        if (head.next == null) return "Reversed"; 
        Node curr = head; 
        while (curr != null) { 
            Node next = curr.next; 
            curr.next = curr.previous; 
            curr.previous = next; 
            curr = next; 
        } 
        Node temp = head;
        head = tail;
        tail = temp;
        return "Reversed"; 
    }
}

class DoublyLinkedList {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        DoublyLinkedListData list = new DoublyLinkedListData();
        int option;
        while (true) {
            System.out.print("\n0. Exit\n1. Insert at Beginning\n2. Delete at Beginning\n3. Insert at Ending\n4. Delete at Ending\n5. Insert after Data\n6. Delete after Data\n7. Insert before Data\n8. Delete before Data\n9. Delete Data\n10. Traverse\n11. Reverse Traverse\n12. Is Empty\n13. Length\n14. Search\n15. Reverse\nOption: ");
            option = sc.nextInt();
            if (option == 0) return;

            switch (option) {
                case 1 -> {
                    System.out.print("Element: ");
                    int element = sc.nextInt();
                    System.out.println(list.insertAtBeginning(element));
                }
                case 2 -> System.out.println(list.deleteAtBeginning());
                case 3 -> {
                    System.out.print("Element: ");
                    int element = sc.nextInt();
                    System.out.println(list.insertAtEnding(element));
                }
                case 4 -> System.out.println(list.deleteAtEnding());
                case 5 -> {
                    System.out.print("Element: ");
                    int element = sc.nextInt();
                    System.out.print("Data: ");
                    int data = sc.nextInt();
                    System.out.println(list.insertAfterData(element, data));
                }
                case 6 -> {
                    System.out.print("Data: ");
                    int data = sc.nextInt();
                    System.out.println(list.deleteAfterData(data));
                }
                case 7 -> {
                    System.out.print("Element: ");
                    int element = sc.nextInt();
                    System.out.print("Data: ");
                    int data = sc.nextInt();
                    System.out.println(list.insertBeforeData(element, data));
                }
                case 8 -> {
                    System.out.print("Data: ");
                    int data = sc.nextInt();
                    System.out.println(list.deleteBeforeData(data));
                }
                case 9 -> {
                    System.out.print("Data: ");
                    int data = sc.nextInt();
                    System.out.println(list.deleteData(data));
                }
                case 10 -> list.traverse();
                case 11 -> list.reverseTraverse();
                case 12 -> System.out.println(list.isEmpty());
                case 13 -> System.out.println(list.getLength());
                case 14 -> {
                    System.out.print("Data: ");
                    int data = sc.nextInt();
                    System.out.println(list.search(data));
                }
                case 15 -> System.out.println(list.reverse());
                default -> System.out.println("Please Choose a valid option");
            }
        }
    }
}

File: Data Structures/Python/doublyLinkedList.py
class Node:
	def __init__(self, element):
		self.data = element
		self.previous = None
		self.next = None

class DoublyLinkedList:
	def __init__(self):
		self.head = None
		self.tail = None

	def insertAtBeginning(self, element):
		newNode = Node(element)
		if self.head == None :
			self.head = newNode
			self.tail = newNode
			return "Inserted"
		newNode.next = self.head
		self.head.previous = newNode
		self.head = newNode
		return "Inserted"

	def deleteAtBeginning(self):
		if self.head == None:
			return "Linked List is Empty"
		if self.head != None and self.head.next == None:
			self.head = None
			self.tail = None
			return "Deleted"
		self.head = self.head.next
		self.head.previous = None
		return "Deleted"

	def insertAtEnding(self, element):
		newNode = Node(element)
		if self.head == None:
			self.head = newNode
			self.tail = newNode
			return "Inserted"
		newNode.previous = self.tail
		self.tail.next = newNode
		self.tail = newNode
		return "Inserted"

	def deleteAtEnding(self):
		if self.head == None:
			return "Linked List is Empty"
		if self.head != None and self.head.next == None:
			self.head = None
			self.tail = None
			return "Deleted"
		self.tail = self.tail.previous
		self.tail.next = None
		return "Deleted"

	def insertAfterData(self, data, element):
		if self.head == None:
			return "Linked List is Empty"
		newNode = Node(element)
		curr = self.head
		while curr:
			if curr.data == data:
				newNode.next = curr.next
				newNode.previous = curr
				if curr.next != None:
					curr.next.previous = newNode
				else:
					self.tail = newNode
				curr.next = newNode
				return "Inserted"
			curr = curr.next
		return "Data no found"
	
	def deleteAfterData(self, data):
		if self.head == None:
			return "Linked List is Empty"
		curr = self.head
		while curr.next != None:
			if curr.data == data:
				if curr.next.next != None:
					curr.next.next.previous = curr
				else:
					self.tail = curr
				curr.next = curr.next.next
				return "Deleted"
			curr = curr.next
		if curr.data == data:
			return "No element after this data"
		return "Data not found"
	
	def insertBeforeData(self, data, element):
		if self.head == None:
			return "Linked List is Empty"
		newNode = Node(element)
		curr = self.head
		if curr.data == data:
			newNode.next = curr
			curr.previous = newNode
			self.head = newNode
			return "Inserted"
		while curr.next != None:
			if curr.next.data == data:
				newNode.next = curr.next
				curr.next.previous = newNode
				newNode.previous = curr
				curr.next = newNode
				return "Inserted"
			curr = curr.next
		return "Data not found"
	
	def deleteBeforeData(self, data):
		if self.head == None:
			return "Linked List is Empty"
		if self.head.data == data:
			return "No element before this data"
		curr = self.head
		while curr.next != None:
			if curr.next.data == data:
				if curr.previous == None:
					self.head = curr.next
				else:
					curr.previous.next = curr.next
				curr.next.previous = curr.previous
				return "Deleted"
			curr = curr.next
		return "Data not found"
	
	def deleteData(self, data):
		if self.head == None:
			return "Linked List is Empty"
		if self.head == self.tail and self.head.data == data:
			self.head = None
			self.tail = None
			return "Deleted"
		curr = self.head
		while curr:
			if curr.data == data:
				if curr.previous == None:
					curr.next.previous = None
					self.head = curr.next
					return "Deleted"
				curr.previous.next = curr.next
				if curr.next != None:
					curr.next.previous = curr.previous
				else:
					self.tail = curr.previous
				return "Deleted"
			curr = curr.next
		return "Data not found"

	def traverse(self):
		if self.head == None:
			print("Linked List is Empty")
			return
		curr = self.head
		while curr:
			print(curr.data, end = " ")
			curr = curr.next
		print("\n")

	def reverseTraverse(self):
		if self.head == None:
			print("Linked List is Empty")
			return
		curr = self.tail
		while curr:
			print(curr.data, end = " ")
			curr = curr.previous
		print("\n")
	
	def isEmpty(self):
		return True if self.head == None else False
	
	def getLength(self):
		i = 0
		curr = self.head
		while curr:
			i += 1
			curr = curr.next
		return i
	
	def search(self, data):
		if self.head == None:
			return "Linked List is Empty"
		curr = self.head
		while curr:
			if curr.data == data:
				return data
			curr = curr.next
		return "Data not found"
	
	def reverse(self):
		if self.head == None:
			return "Linked List is Empty"
		if self.head != None and self.head.next == None:
			return "Reversed"
		prev = None
		curr = self.head
		while curr:
			nsx = curr.next
			curr.next = prev
			curr.previous = nsx
			prev = curr
			curr = nsx
		self.tail = self.head
		self.head = prev
		return "Reversed"

linkedList = DoublyLinkedList()
while 1:
	print("\n0. Exit\n1. Insert at Beginning\n2. Delete at Beginning\n3. Insert at Ending\n4. Delete at Ending\n5. Insert after Data\n6. Delete after Data\n7. Insert before Data\n8. Delete before Data\n9. Delete Data\n10. Traverse\n11. Reverse Traverse\n12. Is Empty\n13. Length\n14. Search\n15. Reverse")
	option = int(input("\nOption: "))
	if option == 0:
		break
	match option:
		case 1:
			element = int(input("\nElement: "))
			print(linkedList.insertAtBeginning(element))
			linkedList.traverse()
		case 2:
			print(linkedList.deleteAtBeginning())
			linkedList.traverse()
		case 3:
			element = int(input("\nElement: "))
			print(linkedList.insertAtEnding(element))
			linkedList.traverse()
		case 4:
			print(linkedList.deleteAtEnding())
			linkedList.traverse()
		case 5:
			element = int(input("\nElement: "))
			data = int(input("\nData: "))
			print(linkedList.insertAfterData(data, element))
			linkedList.traverse()
		case 6:
			data = int(input("\nData: "))
			print(linkedList.deleteAfterData(data))
			linkedList.traverse()
		case 7:
			element = int(input("\nElement: "))
			data = int(input("\nData: "))
			print(linkedList.insertBeforeData(data, element))
			linkedList.traverse()
		case 8:
			data = int(input("\nData: "))
			print(linkedList.deleteBeforeData(data))
			linkedList.traverse()
		case 9:
			data = int(input("\nData: "))
			print(linkedList.deleteData(data))
			linkedList.traverse()
		case 10:
			linkedList.traverse()
		case 11:
			linkedList.reverseTraverse()
		case 12:
			print(linkedList.isEmpty())
		case 13:
			print(linkedList.getLength())
		case 14:
			data = int(input("\nData: "))
			print(linkedList.search(data))
		case 15:
			print(linkedList.reverse())
			linkedList.traverse()
